### PR TITLE
[dhcp_relay]  refactor restart_dhcp_service to explicitly target the active relay agent

### PR DESCRIPTION
**Pull request by Xichen96**
### Description of PR

Reworks restart_dhcp_service in tests/common/dhcp_relay_utils.py so DHCP relay readiness is checked against the specific relay agent the test actually
exercises, instead of the long-standing grep dhc | awk '{print $2}' heuristic (and its later grep -v dhcpmon patch) that asserts every supervisord entry
whose name contains dhc is RUNNING.

The original check has two structural problems that have only become more painful as the relay container grew new modes:

 1. It is implicit. Whether the assertion is correct depends entirely on which entries supervisord happens to expose at that moment, not on what the test cares about. The introduction of has_sonic_dhcpv4_relay mode (single dhcp4relay proc, no per-VLAN procs) already required a one-off patch (grep -v dhcpmon) to keep it working, and a similar patch would be needed every time a new agent variant is added.
 2. There is no readiness API for callers that have already triggered a restart via other means (config load_minigraph, GCU patch apply, VLAN add/remove). Tests that need to wait without restarting again have been growing private polling helpers (e.g. is_dhcrelay_forwarder_running in the mx internal-mode test).

This PR replaces the implicit "everything matching dhc is RUNNING" rule with an explicit, named contract — the caller declares which relay agent variants it expects (isc, isc-internal, sonic, v6) — and exposes the wait-only half of the function for callers that already restarted the service themselves.

Summary:
Fixes # (issue)

### Type of change

- [x] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] New Test case
    - [ ] Skipped for non-supported platforms
- [ ] Test case improvement

### Back port request
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411
- [ ] 202505
- [ ] 202511
- [x] 202605

### Tested branch
- [x] master
- [x] 202605

### Test result
- master: 35 passed, 1 failed — the single failure (`test_dhcp_relay_restart_with_stress[sonic-relay-agent]`) is a pre-existing `dhcp4relay` flap under the test's own DHCP flood; it reproduces without this change (see "How did you verify/test it?" below).
- 202605: `SONiC.20260510.06` — full `tests/dhcp_relay` module on `testbed-bjw2-can-720dt-5` (720dt, `m0`). The new `restart_dhcp_service(duthost, relay_types)` (required/validated arg) and `wait_dhcp_relay_ready(duthost, relay_types)` executed cleanly on hardware across all three mappings — `['isc']`, `['sonic']`, `['v6']` — with no `TypeError` / missing-argument / validation errors; control-plane cases (interface binding, counter/stress, packet-recv) passed. A subset of packet-forwarding cases failed, traced to a local testbed uplink/LACP infrastructure issue (several uplink PortChannels not forming on the fanout) and unrelated to this test-only change — the same cases fail on the unmodified baseline. Evidence: https://github.com/sonic-net/sonic-mgmt/pull/24403#issuecomment-4998944761

#### What is the motivation for this PR?

The historical readiness check assumes a specific supervisord layout (per-VLAN isc-dhcpv4-relay-VlanX + dhcprelayd + dhcp6relay + dhcpmon-VlanX) andasserts they are all RUNNING. As soon as the container layout changed:

 - has_sonic_dhcpv4_relay=True mode replaces N per-VLAN isc-dhcpv4-relay-VlanX procs with a single consolidated dhcp4relay and intentionally stops dhcpmon. The original grep dhc would forever see at least one STOPPED line (dhcpmon-VlanX) and time out. A grep -v dhcpmon patch was added to keep it limping along, but it is just paper over a broken model.
 - mx internal-only mode (dhcprelayd consolidating to dhcrelay -iu docker0 ...) leaves the per-VLAN supervisord entries STOPPED by design, so the same heuristic cannot be reused at all — the mx test had to grow its own is_dhcrelay_forwarder_running poller.

The right fix is to stop guessing and let each call site say what it expects.

#### How did you do it?

 - restart_dhcp_service(duthost, relay_types) now takes a required, explicit iterable. Each entry must be one of 'isc', 'isc-internal', 'sonic', 'v6'. 
 - The orchestrator (dhcprelayd) is always required RUNNING. There is no default — every caller declares its own contract.
 - New wait_dhcp_relay_ready(duthost, relay_types) exposes the wait-only path for callers that already triggered a restart by other means (GCU apply, config load_minigraph, VLAN config edits).
 - Updated all 12 call sites under tests/dhcp_relay/, tests/generic_config_updater/, and the inner call inside sonic_dhcpv4_flag_config_and_unconfig to pass the appropriate relay_types based on the test's parametrization.
 - 'isc-internal' has no public consumer in sonic-mgmt (no tests/mx); it is included so the common utility is complete and the parallel mx test in sonic-mgmt-int can rely on it without further utils churn.

#### How did you verify/test it?

 - bjw2-can-t0-7260-8 (Arista 7260, master image): full dhcp_relay suite — tests/dhcp_relay/test_dhcp_relay.py, test_dhcpv6_relay.py, 
test_dhcp_relay_stress.py, tests/generic_config_updater/test_dhcp_relay.py — 35 passed, 1 failed. The single failure 
(test_dhcp_relay_restart_with_stress[sonic-relay-agent]) is dhcp4relay flapping under the test's own DHCP packet flood (76 STOPPED / 2 STARTING / 20 
RUNNING samples across the 120s wait); reproduces with the pre-refactor check too, tracked separately.
 - mx coverage validated using the parallel internal mx test on bjw2-can-720dt-2 (Mellanox 2700) and bjw2-can-7215-4 (Marvell armhf), 4/4 passed each.

##### Work item tracking
- Microsoft ADO (number only): 38699914

### CODE REVIEW

**Comment by Blueve** [comment]
Thanks for taking on this rework — the structural critique in the PR description is on point. Before LGTM I'd like to see the following addressed; happy to discuss any of them.

### 1. `v4_vlans` discovery is too broad — can deadlock the readiness wait
```python
for vlan, attrs in json.loads(vlan_json).items():
    if attrs.get('dhcp_servers'):
        v4_vlans.append(vlan)
```
But `docker-dhcp-relay.supervisord.conf.j2` + `dhcpv4-relay.agents.j2` only emit `[program:isc-dhcpv4-relay-<vlan>]` when **all three** hold:
- `vlan in VLAN_INTERFACE`
- the VLAN appears in the outer `VLAN_INTERFACE | pfx_filter` loop with an **IPv4** prefix
- at least one entry in `VLAN[vlan].dhcp_servers` passes `dhcp_server | ipv4`

A VLAN with `dhcp_servers` populated but no IPv4 interface (or only v6 entries in `dhcp_servers`) will cause the readiness check to wait forever on a non-existent supervisord entry → `wait_until` times out at 120s. Safer: intersect with `VLAN_INTERFACE` and filter by `ipv4`, or — better — drive the assertion from supervisord's actual layout (require every `isc-dhcpv4-relay-*` already present in `_supervisor_status_map()` to be `RUNNING`) and stop trying to predict it from CONFIG_DB.

### 2. `'isc'` with empty `v4_vlans` silently passes
```python
if 'isc' in relay_types and v4_vlans:
```
If a caller asks for `'isc'` but no v4 vlans were discovered (config drift, parametrize mistake, or bug #1 above), the readiness check collapses to "just `dhcprelayd` is RUNNING" — trivially true — and the test proceeds against a relay container with no v4 service. That's worse than the old failure mode because it fails *silently*. At minimum log a warning when `'isc'` was requested but `v4_vlans` is empty; arguably treat it as a failed precondition.

### 3. Mutually exclusive `relay_types` combinations are accepted
`'sonic'`, `'isc'`, and `'isc-internal'` are mutually exclusive at the supervisord-template level (driven by `has_sonic_dhcpv4_relay` + mx mode). A caller passing e.g. `['isc', 'sonic']` will wait forever because only one of `dhcp4relay` / `isc-dhcpv4-relay-*` ever exists. Either validate the combination up front, or auto-derive the v4 mode from `has_sonic_dhcpv4_relay` (like the old code did) and accept just one v4 token from callers. As a related guard: if a caller passes `['sonic']` against a device whose conf was generated with `has_sonic_dhcpv4_relay=False`, there's no `dhcp4relay` entry and the wait silently times out — a quick `sonic-db-cli hget DEVICE_METADATA|localhost has_sonic_dhcpv4_relay` cross-check with a clear error message would save a lot of debug time on contract mismatches.

### 4. `wait_dhcp_relay_ready` ships with no callers
The PR rationale ("private pollers like `is_dhcrelay_forwarder_running` in the mx internal-mode test") explicitly motivates the new function, but those private helpers aren't replaced here, so the new code path isn't exercised by any test in this PR. Either replace at least one caller in this PR (so CI covers it) or land it in the same change as the migration. As-is, the next person to use it discovers any subtle bugs cold.

### 5. Failure message lacks the actual supervisor state snapshot
```python
"dhcp_relay is not ready (relay_types=%s v4_vlans=%s)" % (relay_types, v4_vlans)
```
When this fires, the immediate next debug step is "what *was* the state?". Capture the last `_supervisor_status_map()` snapshot in the closure and include it in the assertion message — it's free and makes triage from a CI log dramatically easier.

---

Other smaller observations (spurious-match risk on `pgrep -af '/usr/sbin/dhcrelay.*-iu docker0'`, `module_ignore_errors=True` swallowing real container failures, breaking signature change with no soft-deprecation window, ~240 docker-execs in worst-case poll) are nits I'm happy to flag inline if you want, but the five above are the ones I think matter for correctness.

**Comment by Xichen96** [comment]
Thanks for the careful review! Pushed v2 addressing all of these:

1. v4_vlans from CONFIG_DB is too broad — Agreed, removed CONFIG_DB-derived v4_vlans discovery entirely. The 'isc' mode is now driven by supervisord at poll time: every isc-dhcpv4-relay-* entry currently present in the dhcp_relay container must be RUNNING. supervisord (built from dhcpv4-relay.agents.j2) is the source of truth for which per-VLAN agents should exist; a VLAN with v6-only dhcp_servers or no IPv4 interface simply won't have an isc-dhcpv4-relay-<Vlan> entry, and we no longer wait for one.

2. Silent pass when v4_vlans is empty — As discussed, this is the intended behavior: if there are no per-VLAN isc helpers configured in the container layout, the 'isc' check is a legitimate no-op. Callers that expect specific helpers should pair the wait with their own config setup. Documented in the docstring.

3. Mutually exclusive combos accepted — Added an upfront guard: at most one of {'isc', 'isc-internal', 'sonic'} may appear in relay_types. Any other combination raises ValueError immediately rather than spinning until timeout.

4. wait_dhcp_relay_ready had no callers — Wired into test_interface_binding in both test_dhcp_relay.py (['isc']) and test_dhcpv6_relay.py (['v6']), right after wait_critical_processes + the check_interface_status poll, so the v4/v6 binding test now blocks on relay readiness before exercising sockets.

5. Failure message lacks state — Closure now captures the last _supervisor_status_map() snapshot and (when applicable) the pgrep output for the internal dhcrelay proc, surfaced in the assertion message: relay_types=... last_supervisor_states={...} isc_internal_procs=[...].

Nit-1 (pgrep regex too loose) — Skipped: the pgrep runs inside the dhcp_relay container where the only dhcrelay process possible is the one we spawn.
Happy to tighten if you'd prefer defense-in-depth.

Nit-2 (module_ignore_errors=True in poll fns) — Removed. wait_until already converts exceptions raised by the poll function into a False iteration, so the manual rc/stderr plumbing was redundant; let duthost.shell raise and let wait_until retry.

Bonus (not from review): dropped poll cadence from (120, 1, 10) to (240, 5, 10) so we're not hitting the DUT with up to ~240 docker exec calls per wait
— now ~48 max, with a more generous overall timeout.

**Comment by bhouse-nexthop** [comment]
Hi @Xichen96, we re-triggered CI on this PR and it's still failing. Could you take a look at the latest run and address the failures? Once CI is green we'll proceed with review. Thanks!

File: tests/common/dhcp_relay_utils.py [hunk 1]
```diff
@@ -20,26 +20,172 @@
...
     duthost.shell('systemctl reset-failed dhcp_relay')
     duthost.shell('systemctl restart dhcp_relay')
     duthost.shell('systemctl reset-failed dhcp_relay')
 
-    # In SONiC dhcp relay agent mode, dhcprelayd stops dhcpmon by design,
-    # so exclude it from the readiness check (dhcprelayd.py refresh_dhcrelay TODO).
-    has_sonic_relay = duthost.shell(
-        'sonic-db-cli CONFIG_DB hget "DEVICE_METADATA|localhost" "has_sonic_dhcpv4_relay"',
-        module_ignore_errors=True)['stdout'].strip() == 'True'
+    wait_dhcp_relay_ready(duthost, relay_types)
```
[the remark this answers is not in the document]
Comment (Xichen96) [reply tests/common/dhcp_relay_utils.py]: Declining this one. `restart_dhcp_service` / `wait_dhcp_relay_ready` are internal test helpers with a fixed, known set of callers, all of which pass validated list literals (`['isc']`, `['sonic']`, `['v6']`, etc.). An invalid or unsatisfiable `relay_types` (e.g. `['isc', 'sonic']`) never comes from any caller, so the "restart then raise" path isn't reachable in practice — and `wait_dhcp_relay_ready` still raises `ValueError` on genuinely invalid types. Reordering/duplicating the validation for inputs that can't occur here isn't warranted.

File: tests/common/dhcp_relay_utils.py [hunk 2]
```diff
@@ -20,26 +20,172 @@
...
+
+    Same `relay_types` contract as `restart_dhcp_service`. Use this when the caller
+    has already restarted dhcp_relay (or applied config that triggers a restart, e.g.
+    `config reload`, `config load_minigraph`, GCU `apply-patch`) and only needs to
+    block on readiness.
+    """
+    if not relay_types:
+        raise ValueError("wait_dhcp_relay_ready: relay_types must be a non-empty iterable")
+    relay_types = list(relay_types)
+    valid = {'isc', 'isc-internal', 'sonic', 'v6'}
```
[the remark this answers is not in the document]
Comment (Xichen96) [reply tests/common/dhcp_relay_utils.py]: Declining — same reasoning. This is an internal helper and every caller passes a list literal (`['isc']`, `['v6']`, etc.); none pass a bare string or a generator, so the truthiness / char-split edge cases don't arise. Leaving the existing check as-is.

**Comment by yxieca** [comment]
**AI review — Changes requested**

### Blocking

1. `restart_dhcp_service` bounces the service before validating `relay_types` (tests/common/dhcp_relay_utils.py, `restart_dhcp_service` lines 98-105). The function body only checks `if not relay_types: raise ValueError(...)` and then immediately runs `systemctl reset-failed dhcp_relay` / `systemctl restart dhcp_relay` before calling `wait_dhcp_relay_ready(duthost, relay_types)`. The membership check (`valid = {'isc', 'isc-internal', 'sonic', 'v6'}`) and the at-most-one-v4 mutual-exclusion check live in `wait_dhcp_relay_ready` and therefore run *after* the restart. Net effect: a caller passing a typo (e.g. `['scon']`) or a mutually-exclusive combo (e.g. `['isc', 'sonic']`) will restart `dhcp_relay` on the DUT and only then raise `ValueError`, disrupting unrelated test state and producing a confusing failure. This also contradicts the function's own docstring, which states these combinations are "rejected up front." Move the membership + v4-mutex validation into `restart_dhcp_service` before the `systemctl` calls, or route both entry points through a single validator that runs first.

---
Reviewed by AI agent on behalf of Ying.

**Comment by Xichen96** [comment]
Good catch, fixed in eb340a1a3a. Extracted a shared `_validate_relay_types()` (non-empty + membership + at-most-one-v4-mode) and call it at the top of **both** `restart_dhcp_service` (before the `systemctl` calls) and `wait_dhcp_relay_ready`. A typo or mutually-exclusive combo now raises before the service is touched, matching the docstring's "rejected up front."

**Comment by Xichen96** [comment]
Gentle ping for re-review 🙏 — the blocking comment (shared relay-agent validator moved up-front) is addressed in eb340a1a3a, and CI is now green. The earlier red run was a `LOCK_TESTBED` infra timeout (2h lock wait, 0 tests executed), which cleared on rerun — not a code failure. @yxieca @StormLiangMS could you take another look when you have a moment? Thanks!

**Review by StormLiangMS — COMMENTED** [review]
@Xichen96 — ✅ Reviewed the latest head; no blocking issues found.

**Template**: ✅ Complete, with ADO tracking and 202605 backport context.
**DCO / CI**: ✅ Green.

The shared `_validate_relay_types()` now rejects invalid input before restarting `dhcp_relay`, and the supervisord-driven readiness checks correctly target the requested active relay layout. The updated callers consistently declare the expected relay type. The remaining empty-generator normalization edge case is non-blocking because all current callers pass list literals.

**Review by StormLiangMS — APPROVED** [review]
LGTM

**Comment by StormLiangMS** [comment]
hi @Xichen96 could you test with 202605 for the cherrypick ask?

**Comment by Xichen96** [comment]
### 202605 hardware validation

Cherry-picked this change onto `internal-202605` and ran the full `tests/dhcp_relay` module on a **202605** image (`SONiC.20260510.06`) on physical hardware (720dt, `m0` topology).

**Result: the refactor works as intended on hardware.** The new `restart_dhcp_service(duthost, relay_types)` signature (required + validated `relay_types`) and the new `wait_dhcp_relay_ready(duthost, relay_types)` helper executed cleanly across all three mappings — `['isc']`, `['sonic']`, and `['v6']` — with no `TypeError` / missing-argument / validation errors. Control-plane cases passed (interface binding, counter/stress, packet-recv).

A subset of packet-**forwarding** cases failed, but those were traced to a local testbed link/LACP infrastructure issue (several uplink PortChannels not forming on the fanout) and are **unrelated to this test-only change** — the same cases fail on the unmodified baseline. No regression attributable to this PR.
